File: March2022/Java/ChampagneTower.java

// @saorav21994
// TC : O(no. of rows^2)
// SC : O(no. of rows^2)


class Solution {
    public double champagneTower(int poured, int query_row, int query_glass) {
        double [][]dp = new double[102][102];
        dp[0][0] = (double) poured;
        for (int r = 0; r <= query_row; ++r) {
            for (int c = 0; c <= r; ++c) {
                double res = (dp[r][c] - 1.0) / 2.0;
                if (res > 0) {
                    dp[r+1][c] += res;
                    dp[r+1][c+1] += res;
                }
            }
        }
        return Math.min(1, dp[query_row][query_glass]);
    }
}

File: March2022/C++/Champagne_Tower.cpp
class Solution {
public:
    double champagneTower(int poured, int query_row, int query_glass) {
        vector<vector<double>> dp(101,vector<double>(101,0));
        dp[0][0]=poured;
        double rest=0;
        for(int i=0;i<query_row;i++)
        {
            for(int j=0;j<101;j++)
            {
                if(dp[i][j]>1)
                {
                    rest=dp[i][j]-1;
                    dp[i+1][j]+=rest/2.0;
                    dp[i+1][j+1]+=rest/2.0;
                }
            }
        }
        return dp[query_row][query_glass]>1?1:dp[query_row][query_glass];
    }
};